### PR TITLE
[internal_medicine] Absorb llm-internal-medicine into paddlefleet

### PR DESCRIPTION
**Pull request by yuweivvv**
### PR Category
Execute Infrastructure

### PR Types
New features, Bug fixes, Docs

## What this PR does

Absorbs the `llm-internal-medicine` monitoring library (previously an
erniebot git submodule pointing at `bo-ke/llm-internal-medicine@dd54aef5`)
into PaddleFleet as `paddlefleet.internal_medicine`.

Motivation: the library is only ever used through the PaddleFleet trainer
callback, so keeping it as a separate submodule meant the monitors and the
trainer that calls them could drift apart across versions. After this PR the
two evolve as one version, and erniebot drops one submodule.

## Where things land

- `src/paddlefleet/internal_medicine/` — the library (core + megatron/paddlefleet backends)
- `docs/internal_medicine/` — the per-family docs, plus a development guide carried over from the source repo's `CLAUDE.md`
- `tests/single_card_tests/internal_medicine/` — the unit tests
- `.agents/skills/monitor-hook-perf-rules/SKILL.md` — the monitor hot-path discipline doc

Integration points touched (2 files, 30 lines total):

- `src/paddlefleet/trainer/trainer_callback.py` — imports switch to the in-tree package
- `src/paddlefleet/cli/hparams/finetuning_args.py` — doc string points at the new module path

The three megatron/torch backend tests are listed in
`tests/single_card_tests/disable_single_card_uts.txt`: the single-card CI image
has no torch, so they run in the megatron env only, selected via
`INTERNAL_MEDICINE_TEST_BACKEND`.

## Equivalence with the source library

The move is content-preserving against source commit `dd54aef5`. Every one of
the 43 Python modules, 13 test files and 12 docs is present, and an AST-level
comparison of each module shows no behavioural change except the follow-up
fixes listed below. Cosmetic differences from the move itself:

- Apache license headers added
- reformatted to this repo's line length
- lint-driven rewrites: `import torch.nn as nn` -> `from torch import nn`,
  `key[len(p):]` -> `key.removeprefix(p)`, `str = None` -> `str | None = None`,
  `collections.abc` imports moved under `TYPE_CHECKING`
- import prefix updated to `paddlefleet.internal_medicine` (code, docs, tests)
- `sys.path.insert(..., "src")` removed from the tests, which is no longer
  needed now that the package is importable

Deliberately **not** included: the source repo's `master` is two commits ahead
of `dd54aef5` with a per-layer activation-gradient monitor
(`bo-ke/llm-internal-medicine#72`). That monitor will follow as its own PR.

## Follow-up fixes in this PR

These were pre-existing in `dd54aef5` and surfaced by review. They are in
this PR rather than a follow-up so the in-tree library does not ship the
known correctness/perf issues.

1. **CP QK reconstruction is layout-aware.** Dual-chunk CP stores a front
   chunk plus a rear chunk on every rank, so `cp_rank * q_local_seq` plus a
   plain `all_gather` silently misaligned causal / entropy / sink. The
   monitor now restores full-sequence Q and K with the matching gather
   (`all_gather_contiguous` vs `all_gather_balance`) and runs the kernel at
   `q_row_offset=0`.
2. **Reducer layout cache is per-rank.** PP/VPP ranks may own disjoint
   metric-key sets, so comparing fingerprints across ranks re-aligned every
   sampled step and reintroduced `all_gather_object`. Each rank now compares
   against its own `_local_fingerprint` and uses a single `MAX(stale)`
   collective to trigger a global re-align only when any rank's schema
   changes.
3. **`IM_DISABLE_REDUCE` removed.** The numeric reducer is the only
   configured distributed path. `all_gather_object` remains an internal
   fallback and is no longer selected by an undeclared environment variable.

## Verification

Unit tests:

- `pytest tests/single_card_tests/internal_medicine` — 342 passed, 5
  subtests passed (3 torch-only files excluded, as CI does)
- Added `test_reducer_keeps_aligned_layout_for_heterogeneous_rank_keys`:
  two reducer instances with disjoint PP-stage keys merge once and do not
  re-align on the next step

Single-node 8-GPU runs:

- ERNIELite MoE, 11 steps, CP=1: seven monitor families collect and flush.
  `internal_medicine.jsonl` has 2523 keys/record, no NaN/Inf.
- MQA+DSA (`hybrid_mla_attention: mqa_dsa`, sparse indexer loss, cuDNN
  indexer, CP=1): `dsa_health` discovers 3 DSA layers, 19 metrics each plus
  the global aggregate.
- PP=2 + EP=4 heterogeneous keys: each PP stage sees 6 attention layers
  locally; the flushed jsonl is the 13-layer union. Layout aligned once
  (`1391 keys`), not on later steps.
- CP=2 dual-chunk vs contiguous QK comparison on `eb5_v2_10B_A1B` (not
  `dsv4_hybrid`, so dual-chunk is legal), 6 steps, `seed=42`,
  `FLAGS_embedding_deterministic=1`, `FLAGS_cudnn_deterministic=1`.
  Both runs complete. QK key sets are identical (416 keys). Global
  entropy/sink match to `1e-5`; the previous silent dual-chunk misalignment
  is gone. Residual `0.2%–0.8%` diffs remain on QKV-norm / logit-max and
  track training-numeric differences between the two CP layouts, not row
  mapping.

## Known issue, pre-existing and not introduced here

`dsa_health/select_iou_prev` is only recorded on one layer under
`recompute_granularity: full`. The hooks fire during the backward recompute
pass, which walks layers in descending order, so the ascending-order guard
in `_record_cross_layer` rejects every same-pass neighbour pair and only
accepts a cross-microbatch wrap-around. This behaviour is identical in the
source library (the file was AST-identical at the move) and is out of scope
here.

## Coordination

The matching erniebot CR (which removes the submodule and advances this
submodule pointer) must land together with this PR, otherwise erniebot
would briefly have the submodule removed while PaddleFleet does not yet
carry the library, and training would silently lose monitoring. The
erniebot pointer tracks this PR head and will be updated to the merged
commit before that CR is submitted.

### CODE REVIEW

File: src/paddlefleet/internal_medicine/backends/paddlefleet/qk_monitor.py [hunk 1]
```diff
@@ -0,0 +1,1083 @@
...
+                                layer_idx,
+                                f"{name}_norm_max",
+                                vector_norm.max(),
+                                attn_type=attn_type,
+                            )
+                    if self.cp_size > 1:
+                        # CP > 1: gather K only, keep Q local.
+                        query = query.detach()
+                        q_local_seq = query.shape[1]  # rows on this rank
+                        q_row_offset = self.cp_rank * q_local_seq
```
[the remark this answers is not in the document]
Comment (yuweivvv) [reply src/paddlefleet/internal_medicine/backends/paddlefleet/qk_monitor.py]: 确认这是实际的正确性问题。已修复：QK monitor 现在读取 `cp_balance_mode`；只有 `contiguous_allgather` 使用本地 Q 的连续偏移和 K gather 路径，默认的 `dualchunk_allgather` 不再执行会错位的 QK/causal/entropy/sink 重建，避免静默产生错误指标。QKV norm 仍保留。已补充静态检查并通过内科测试：`342 passed, 5 subtests passed`。

File: src/paddlefleet/internal_medicine/backends/paddlefleet/gather.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    # ------------------------------------------------------------------
+    # Layout agreement
+    # ------------------------------------------------------------------
+
+    def _agree_on_layout(self, schema, paddle, dist) -> bool:
+        """True when every rank still shares the cached layout."""
+        local_fp = schema.fingerprint
+        stale = 1 if (not self._layout or local_fp != self._fingerprint) else 0
+        probe = paddle.to_tensor([local_fp, stale], dtype="int64")
+        dist.all_reduce(probe, op=dist.ReduceOp.MAX)
```
[the remark this answers is not in the document]
Comment (yuweivvv) [reply src/paddlefleet/internal_medicine/backends/paddlefleet/gather.py:80]: 确认。已修复 reducer 的布局缓存：不再比较不同 PP/VPP rank 之间必然可能不同的 fingerprint，改为每个 rank 只和自己的 `_local_fingerprint` 比较，再用一次 `MAX(stale)` collective 触发全局重新对齐。固定 schema 时不会每步重新执行 `all_gather_object`；只有本地 schema 变化时才会对齐。已通过 `342 passed, 5 subtests passed`。

File: src/paddlefleet/internal_medicine/backends/paddlefleet/gather.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    import os
+
+    from ...core.training_logs import training_logs
+
+    try:
+        import paddle.distributed as dist
+
+        if dist.is_initialized():
+            training_logs.set_gather_fn(_paddle_gather)
+            if os.environ.get("IM_DISABLE_REDUCE", "") == "1":
```
[the remark this answers is not in the document]
Comment (yuweivvv) [reply src/paddlefleet/internal_medicine/backends/paddlefleet/gather.py]: 确认。已移除 `IM_DISABLE_REDUCE` 环境变量旁路，numeric reducer 现在是唯一配置化的默认分布式聚合路径；`all_gather_object` 仅作为内部 fallback 保留，不再能通过未声明的环境变量切换通信和性能行为。README 中对应的旧说明也已更新。

**Review by risemeup1111 — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]

**Review by From00 — APPROVED** [review]
LGTM
